### ISSUE TITLE
4D integration discussion (v2.0.0)

### ISSUE THREAD

**Issue by wlthomson** [openmsupply/unified-codes#92]
## Description

Epic for planning and tracking mSupply integration tasks for `v2.0.0`.

Note: this epic only relates to integrating mSupply with new (`>= v2`) functionality, for backwards compatibility with previous (`v1`) functionality, see #9.

## Initiative / goal

Integrate mSupply with basic `v2.0.0` functionality.

## Scope

TODO.

## Issues

TODO.

**Comment by wlthomson** [comment]
Some brief background notes and thoughts. Will keep it brief for now, will add more details if/when needed!

Scope:

- Integrate mSupply with `v2` data service, including new properties, fields, and external data.

Background:

- Existing integration with `v1` API, search for items by code/name.
- Graph entity code/name data can be exposed to existing mSupply instances by "mocking" the `v1` API and mapping `v2` data to legacy format (see #9).
- Legacy API is only for exposing `v1` functionality, should not be extended to expose `v2` functionality (team agree this would be "mission creep" and would cause undesirable dependencies/coupling).
- 4D has no built-in support for GraphQL queries or suitable plugins (@wlthomson has worked on a proof-of-concept demo of a GraphQL plugin using C++ parser libraries).

Next steps:

- IMO, before we can move ahead on this, the team needs to agree on one of the two architectures (open to any other suggestions!):

1. Restrict data service to GraphQL, access `v2` data directly by implementing GraphQL schema/query logic in 4D.
2. Update data service to support both GraphQL and a `v2` RESTful API.

@mark-prins @katherine-sussol Thoughts?

**Comment by mark-prins** [comment]
I prefer keeping the API clean and pushing the complexity into a graphQL client in mSupply (i.e. option 1). That is dependent on how long it would take though and if it is overly complex to do, then sure, some parts of the API could be surfaced via REST. It is only a limited subset of functionality which is required in mSupply:
- I have a universal code, give me interactions
- I have a universal code, give me benchmark pricing
So the overhead of replicating functionality with a REST API is very limited ( at this stage of understanding, at least )

**Comment by wlthomson** [comment]
@mark-prins This is pretty much inline with what I'm thinking. The added benefit of a GraphQL client in mSupply/4D is that it can be recycled for other projects, so my feeling is that it may be worth the time investment, the majority of which is just writing some C++ code and linking with one of the GraphQL libraries (I've already got C++ plugins working, which I think is probably the tricky part).

**Comment by katherine-sussol** [comment]
Yeah, I agree with the above 2 comments, graphQL client would be the ideal solution. 
Personally, I think if we need to go down the REST path it's not terrible to add it to the legacy api - the legacy api is kind of already there just for mSupply support (and also think having 3 different entry points is a little much for such a small set of requests). But - having it in the data-service is definitely cleaner from a modularity point of view!

**Comment by wlthomson** [comment]
@katherine-sussol I think there is a distinction to be made between the legacy API and mSupply compatibility, in that the legacy API is only intended to provide backwards compatibility for existing mSupply instances (i.e. it can be deprecated once mSupply is updated). That's the main reason why I think it would be mission creep to extend the legacy API to other functionality, as I think we want to avoid having to maintain a RESTful API only for mSupply :(.
